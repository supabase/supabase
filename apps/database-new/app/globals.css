@tailwind base;
@tailwind components;
@tailwind utilities;

@import '~ui/build/css/source/global.css';
@import '~ui/build/css/themes/dark.css';
@import '~ui/build/css/themes/light.css';

@font-face {
  font-family: 'circular';
  src: url(/fonts/custom/CustomFont-Book.woff2) format('woff2'),
    url(/fonts/custom/CustomFont-Book.woff) format('woff');
  font-weight: 400;
  font-style: normal;
}
@font-face {
  font-family: 'circular';
  src: url(/fonts/custom/CustomFont-BookItalic.woff2) format('woff2'),
    url(/fonts/custom/CustomFont-BookItalic.woff) format('woff');
  font-weight: 400;
  font-style: italic;
}
@font-face {
  font-family: 'circular';
  src: url(/fonts/custom/CustomFont-Medium.woff2) format('woff2'),
    url(/fonts/custom/CustomFont-Medium.woff) format('woff');
  font-weight: 500;
  font-style: normal;
}
@font-face {
  font-family: 'circular';
  src: url(/fonts/custom/CustomFont-MediumItalic.woff2) format('woff2'),
    url(/fonts/custom/CustomFont-MediumItalic.woff) format('woff');
  font-weight: 500;
  font-style: italic;
}
@font-face {
  font-family: 'circular';
  src: url(/fonts/custom/CustomFont-Bold.woff2) format('woff2'),
    url(/fonts/custom/CustomFont-Bold.woff) format('woff');
  font-weight: 700;
  font-style: 600;
}
@font-face {
  font-family: 'circular';
  src: url(/fonts/custom/CustomFont-BoldItalic.woff2) format('woff2'),
    url(/fonts/custom/CustomFont-BoldItalic.woff) format('woff');
  font-style: 600;
  font-style: italic;
}
@font-face {
  font-family: 'circular';
  src: url(/fonts/custom/CustomFont-Black.woff2) format('woff2'),
    url(/fonts/custom/CustomFont-Black.woff) format('woff');
  font-weight: 800;
  font-style: normal;
}
@font-face {
  font-family: 'circular';
  src: url(/fonts/custom/CustomFont-BlackItalic.woff2) format('woff2'),
    url(/fonts/custom/CustomFont-BlackItalic.woff) format('woff');
  font-weight: 800;
  font-style: italic;
}

@font-face {
  font-family: 'source code pro';
  src: url('/fonts/source-code-pro/SourceCodePro-Regular.eot');
  src: url('/fonts/source-code-pro/SourceCodePro-Regular.woff2') format('woff2'),
    url('/fonts/source-code-pro/SourceCodePro-Regular.woff') format('woff'),
    url('/fonts/source-code-pro/SourceCodePro-Regular.ttf') format('truetype'),
    url('/fonts/source-code-pro/SourceCodePro-Regular.svg#SourceCodePro-Regular') format('svg');
  font-weight: normal;
  font-style: normal;
  font-display: swap;
}

html,
body,
main {
  /* @apply bg-background;
  @apply text-foreground;
  height: 100%;
  width: 100%; */
  padding: 0;
  margin: 0;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.chat-shimmering-loader {
  animation: shimmer 1.5s infinite linear;
  background: linear-gradient(
    to right,
    hsl(var(--background-default)) 0%,
    hsl(var(--brand-default)) 25%,
    hsl(var(--brand-300)) 35%,
    hsl(var(--background-default)) 45%,
    hsl(var(--background-surface-100)) 75%
  );
  background-size: 3000px 100%;
}

.shimmering-loader {
  animation: shimmer 2s infinite linear;
  background: linear-gradient(
    to right,
    hsl(var(--border-default)) 4%,
    hsl(var(--background-surface-200)) 25%,
    hsl(var(--border-default)) 36%
  );
  background-size: 1000px 100%;
}

.dark .shimmering-loader {
  animation: shimmer 2s infinite linear;
  background: linear-gradient(
    to right,
    hsl(var(--border-default)) 4%,
    hsl(var(--border-control)) 25%,
    hsl(var(--border-default)) 36%
  );
  background-size: 1000px 100%;
}

@keyframes shimmer {
  0% {
    background-position: -1000px 0;
  }
  100% {
    background-position: 1000px 0;
  }
}

.monaco-editor,
.monaco-diff-editor {
  --vscode-editor-background: hsl(var(--background-alternative)) !important;
  --vscode-editorGutter-background: hsl(var(--background-alternative)) !important;
  padding-top: 12px;
}

/* .monaco-editor .overflow-guard  */

.shimmering-loader {
  animation: shimmer 2s infinite linear;
  background: linear-gradient(
    to right,
    hsl(var(--border-default)) 4%,
    hsl(var(--background-surface-100)) 25%,
    hsl(var(--border-default)) 36%
  );
  background-size: 1000px 100%;
}

.dark .shimmering-loader {
  animation: shimmer 2s infinite linear;
  background: linear-gradient(
    to right,
    hsl(var(--background-surface-100)) 4%,
    hsl(var(--border-default)) 25%,
    hsl(var(--background-surface-100)) 36%
  );
  background-size: 1000px 100%;
}

@keyframes shimmer {
  0% {
    background-position: -1000px 0;
  }
  100% {
    background-position: 1000px 0;
  }
}
