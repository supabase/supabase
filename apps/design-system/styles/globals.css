@tailwind base;
@tailwind components;
@tailwind utilities;

@import './../../../packages/ui/build/css/source/global.css';
@import './../../../packages/ui/build/css/themes/dark.css';
@import './../../../packages/ui/build/css/themes/classic-dark.css';
@import './../../../packages/ui/build/css/themes/light.css';

@layer base {
  :root {
    --chart-1: 12 76% 61%;
    --chart-2: 173 58% 39%;
    --chart-3: 197 37% 24%;
    --chart-4: 43 74% 66%;
    --chart-5: 27 87% 67%;
  }

  .dark {
    --chart-1: 220 70% 50%;
    --chart-2: 160 60% 45%;
    --chart-3: 30 80% 55%;
    --chart-4: 280 65% 60%;
    --chart-5: 340 75% 55%;
  }
}

@layer base {
  * {
    @apply border-border;
  }
  html {
    @apply scroll-smooth;
  }
  body {
    @apply bg text-foreground;
    /* font-feature-settings: "rlig" 1, "calt" 1; */
    font-synthesis-weight: none;
    text-rendering: optimizeLegibility;
  }
}

@layer utilities {
  .step {
    counter-increment: step;
  }

  .step:before {
    @apply absolute w-9 h-9 bg-muted rounded-full font-mono font-medium text-center text-base inline-flex items-center justify-center -indent-px border-4 border-background;
    @apply ml-[-50px] mt-[-4px];
    content: counter(step);
  }

  .chunk-container {
    @apply shadow-none;
  }

  .chunk-container::after {
    content: '';
    @apply absolute -inset-4 shadow-xl rounded-xl border;
  }
}

@media (max-width: 640px) {
  .container {
    @apply px-4;
  }
}
