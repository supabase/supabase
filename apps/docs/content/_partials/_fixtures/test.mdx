Here is some content to embed as a partial.

## A heading

Some more stuff, including:

- A list
- With two items

And some custom components:

<Admonition type="note">

Some stuff.

In two paragraphs.

</Admonition>

<Component />

An entire paragraph: It was the best of times, it was the worst of times, it was the age of wisdom, it was the age of foolishness, it was the epoch of belief, it was the epoch of incredulity, it was the season of Light, it was the season of Darkness, it was the spring of hope, it was the winter of despair, we had everything before us, we had nothing before us, we were all going direct to Heaven, we were all going direct the other way...
