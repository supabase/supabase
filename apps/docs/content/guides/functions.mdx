---
id: 'functions'
title: 'Edge Functions'
description: 'Globally distributed TypeScript functions.'
subtitle: 'Globally distributed TypeScript functions.'
sidebar_label: 'Overview'
tocVideo: 'za_loEtS4gs'
---

Edge Functions are server-side TypeScript functions, distributed globally at the edge—close to your users. They can be used for listening to webhooks or integrating your Supabase project with third-parties [like Stripe](https://github.com/supabase/supabase/tree/master/examples/edge-functions/supabase/functions/stripe-webhooks). Edge Functions are developed using [Deno](https://deno.com), which offers a few benefits to you as a developer:

- It is open source.
- It is portable. Supabase Edge Functions run locally, and on any other Deno-compatible platform (including self-hosted infrastructure).
- It is TypeScript first and supports WASM.
- Edge Functions are globally distributed for low-latency.

## How it works

- **Request enters an edge gateway (relay)** — the gateway routes traffic, handles auth headers/JWT validation, and applies routing/traffic rules.
- **Auth & policies are applied** — the gateway (or your function) can validate Supabase JWTs, apply rate-limits, and centralize security checks before executing code.
- **[Edge runtime](https://github.com/supabase/edge-runtime) executes your function** — the function runs on a regionally-distributed Edge Runtime node closest to the user for minimal latency.
- **Integrations & data access** — functions commonly call Supabase APIs (Auth, Postgres, Storage) or third-party APIs. For Postgres, prefer connection strategies suited for edge/serverless environments (see the `connect-to-postgres` guide).
- **Observability and logs** — invocations emit logs and metrics you can explore in the dashboard or downstream monitoring (Sentry, etc.).
- **Response returns via the gateway** — the gateway forwards the response back to the client and records request metadata.

## Quick technical notes

- **Runtime:** Supabase Edge Runtime (Deno compatible runtime with TypeScript first). Functions are simple `.ts` files that export a handler.
- **Local dev parity:** Use Supabase CLI for a local runtime similar to production for faster iteration (`supabase functions serve` command).
- **Global deployment:** Deploy your Edge Functions via Supabase Dashboard, CLI or MCP.
- **Cold starts & concurrency:** cold starts are possible — design for short-lived, idempotent operations. Heavy long-running jobs should be moved to [background workers](/docs/guides/functions/background-tasks).
- **Database connections:** treat Postgres like a remote, pooled service — use connection pools or serverless-friendly drivers.
- **Secrets:** store credentials in Supabase [project secrets](/docs/reference/cli/supabase-secrets) and access them via environment variables.

## When to use Edge Functions

- Authenticated or public HTTP endpoints that need low latency.
- Webhook receivers (Stripe, GitHub, etc.).
- On-demand image or Open Graph generation.
- Small AI inference tasks or orchestrating calls to external LLM APIs (like OpenAI)
- Sending transactional emails.
- Building messaging bots for Slack, Discord, etc.

<div className="not-prose">
  <Button size="medium" asChild>
    <a href="/docs/guides/functions/quickstart">Get started</a>
  </Button>
</div>

## Examples

Check out the [Edge Function Examples](https://github.com/supabase/supabase/tree/master/examples/edge-functions) in our GitHub repository.

<div className="grid md:grid-cols-12 gap-4 not-prose">
  {[
    {
      name: 'With supabase-js',
      description: 'Use the Supabase client inside your Edge Function.',
      href: '/guides/functions/auth',
    },
    {
      name: 'Type-Safe SQL with Kysely',
      description:
        'Combining Kysely with Deno Postgres gives you a convenient developer experience for interacting directly with your Postgres database.',
      href: '/guides/functions/kysely-postgres',
    },
    {
      name: 'Monitoring with Sentry',
      description: 'Monitor Edge Functions with the Sentry Deno SDK.',
      href: '/guides/functions/examples/sentry-monitoring',
    },
    {
      name: 'With CORS headers',
      description: 'Send CORS headers for invoking from the browser.',
      href: '/guides/functions/cors',
    },
    {
      name: 'React Native with Stripe',
      description: 'Full example for using Supabase and Stripe, with Expo.',
      href: 'https://github.com/supabase-community/expo-stripe-payments-with-supabase-functions',
    },
    {
      name: 'Flutter with Stripe',
      description: 'Full example for using Supabase and Stripe, with Flutter.',
      href: 'https://github.com/supabase-community/flutter-stripe-payments-with-supabase-functions',
    },
    {
      name: 'Building a RESTful Service API',
      description:
        'Learn how to use HTTP methods and paths to build a RESTful service for managing tasks.',
      href: 'https://github.com/supabase/supabase/blob/master/examples/edge-functions/supabase/functions/restful-tasks/index.ts',
    },
    {
      name: 'Working with Supabase Storage',
      description: 'An example on reading a file from Supabase Storage.',
      href: 'https://github.com/supabase/supabase/blob/master/examples/edge-functions/supabase/functions/read-storage/index.ts',
    },
    {
      name: 'Open Graph Image Generation',
      description: 'Generate Open Graph images with Deno and Supabase Edge Functions.',
      href: '/guides/functions/examples/og-image',
    },
    {
      name: 'OG Image Generation & Storage CDN Caching',
      description: 'Cache generated images with Supabase Storage CDN.',
      href: 'https://github.com/supabase/supabase/tree/master/examples/edge-functions/supabase/functions/og-image-with-storage-cdn',
    },
    {
      name: 'Get User Location',
      description: `Get user location data from user's IP address.`,
      href: 'https://github.com/supabase/supabase/tree/master/examples/edge-functions/supabase/functions/location',
    },
    {
      name: 'Cloudflare Turnstile',
      description: `Protecting Forms with Cloudflare Turnstile.`,
      href: '/guides/functions/examples/cloudflare-turnstile',
    },
    {
      name: 'Connect to Postgres',
      description: `Connecting to Postgres from Edge Functions.`,
      href: '/guides/functions/connect-to-postgres',
    },
    {
      name: 'GitHub Actions',
      description: `Deploying Edge Functions with GitHub Actions.`,
      href: '/guides/functions/examples/github-actions',
    },
    {
      name: 'Oak Server Middleware',
      description: `Request Routing with Oak server middleware.`,
      href: 'https://github.com/supabase/supabase/tree/master/examples/edge-functions/supabase/functions/oak-server',
    },
    {
      name: 'Hugging Face',
      description: `Access 100,000+ Machine Learning models.`,
      href: '/guides/ai/examples/huggingface-image-captioning',
    },
    {
      name: 'Amazon Bedrock',
      description: `Amazon Bedrock Image Generator`,
      href: '/guides/functions/examples/amazon-bedrock-image-generator',
    },
    {
      name: 'OpenAI',
      description: `Using OpenAI in Edge Functions.`,
      href: '/guides/ai/examples/openai',
    },
    {
      name: 'Stripe Webhooks',
      description: `Handling signed Stripe Webhooks with Edge Functions.`,
      href: '/guides/functions/examples/stripe-webhooks',
    },
    {
      name: 'Send emails',
      description: `Send emails in Edge Functions with Resend.`,
      href: '/guides/functions/examples/send-emails',
    },
    {
      name: 'Web Stream',
      description: `Server-Sent Events in Edge Functions.`,
      href: 'https://github.com/supabase/supabase/tree/master/examples/edge-functions/supabase/functions/streams',
    },
    {
      name: 'Puppeteer',
      description: `Generate screenshots with Puppeteer.`,
      href: '/guides/functions/examples/screenshots',
    },
    {
      name: 'Discord Bot',
      description: `Building a Slash Command Discord Bot with Edge Functions.`,
      href: '/guides/functions/examples/discord-bot',
    },
    {
      name: 'Telegram Bot',
      description: `Building a Telegram Bot with Edge Functions.`,
      href: '/guides/functions/examples/telegram-bot',
    },
    {
      name: 'Upload File',
      description: `Process multipart/form-data.`,
      href: 'https://github.com/supabase/supabase/tree/master/examples/edge-functions/supabase/functions/file-upload-storage',
    },
    {
      name: 'Upstash Redis',
      description: `Build an Edge Functions Counter with Upstash Redis.`,
      href: '/guides/functions/examples/upstash-redis',
    },
    {
      name: 'Rate Limiting',
      description: `Rate Limiting Edge Functions with Upstash Redis.`,
      href: '/guides/functions/examples/rate-limiting',
    },
    {
      name: 'Slack Bot Mention Edge Function',
      description: `Slack Bot handling Slack mentions in Edge Function`,
      href: '/guides/functions/examples/slack-bot-mention',
    },
  ].map((x) => (
    <div className="col-span-4" key={x.href}>
      <Link href={x.href} passHref>
        <GlassPanel icon={'/docs/img/icons/github-icon'} hasLightIcon={true} title={x.name}>
          {x.description}
        </GlassPanel>
      </Link>
    </div>
  ))}
</div>
