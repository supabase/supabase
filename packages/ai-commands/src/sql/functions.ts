import { SchemaBuilder } from '@serafin/schema-builder'
import { codeBlock, stripIndent } from 'common-tags'
import { jsonrepair } from 'jsonrepair'
import OpenAI from 'openai'
import { ContextLengthError, EmptyResponseError, EmptySqlError } from '../errors'

const debugSqlSchema = SchemaBuilder.emptySchema()
  .addString('solution', {
    description: 'A short suggested solution for the error (as concise as possible).',
  })
  .addString('sql', {
    description:
      'The SQL rewritten to apply the solution. Includes all the original logic, but modified to fix the issue.',
  })

const generateTitleSchema = SchemaBuilder.emptySchema()
  .addString('title', {
    description: stripIndent`
      The generated title for the SQL snippet (short and concise).
      - Omit these words: 'SQL', 'Postgres', 'Query', 'Database'
    `,
  })
  .addString('description', {
    description: stripIndent`
      The generated description for the SQL snippet.
    `,
  })

// Reference auto-generated types for each JSON schema
export type DebugSqlResult = typeof debugSqlSchema.T
export type GenerateTitleResult = typeof generateTitleSchema.T

// Combine the completion functions
const completionFunctions = {
  debugSql: {
    name: 'debugSql',
    description: stripIndent`
      Debugs a Postgres SQL error. Returns the fixed SQL and a solution explaining it.
      - Create extensions if they are missing (only for valid extensions)
      - Suggest creating tables if they are missing
      - Include all of the original SQL
      - For primary keys, always use "id bigint primary key generated always as identity" (not serial)
      - When creating tables, always add foreign key references inline
      - Prefer 'text' over 'varchar'
      - Prefer 'timestamp with time zone' over 'date'
      - Use vector(384) data type for any embedding/vector related query
      - Always use double apostrophe in SQL strings (eg. 'Night''s watch')
      - Always use semicolons
    `,
    parameters: debugSqlSchema.schema as Record<string, unknown>,
  },
  generateTitle: {
    name: 'generateTitle',
    description: stripIndent`
      Generates a short title and summarized description for a Postgres SQL snippet.
      
      The description should describe why this table was created (eg. "Table to track todos")
    `,
    parameters: generateTitleSchema.schema as Record<string, unknown>,
  },
} satisfies Record<string, OpenAI.Chat.Completions.ChatCompletionCreateParams.Function>

/**
 * Debugs SQL errors.
 *
 * @returns A suggested SQL fix along with a solution explanation.
 */
export async function debugSql(
  openai: OpenAI,
  errorMessage: string,
  sql: string,
  entityDefinitions?: string[]
) {
  const hasEntityDefinitions = entityDefinitions !== undefined && entityDefinitions.length > 0

  const completionMessages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = []

  if (hasEntityDefinitions) {
    completionMessages.push({
      role: 'user',
      content: codeBlock`
        Here is my database schema for reference:
        ${entityDefinitions.join('\n\n')}
      `,
    })
  }

  completionMessages.push(
    {
      role: 'user',
      content: stripIndent`
        Here is my current SQL:
        ${sql}
      `,
    },
    {
      role: 'user',
      content: stripIndent`
        Here is the error I am getting:
        ${errorMessage}
      `,
    }
  )

  try {
    const completionResponse = await openai.chat.completions.create({
      model: 'gpt-3.5-turbo-0125',
      messages: completionMessages,
      max_tokens: 2048,
      temperature: 0,
      tool_choice: {
        type: 'function',
        function: {
          name: completionFunctions.debugSql.name,
        },
      },
      tools: [
        {
          type: 'function',
          function: completionFunctions.debugSql,
        },
      ],
      stream: false,
    })

    const [firstChoice] = completionResponse.choices
    const [firstTool] = firstChoice.message?.tool_calls ?? []

    const sqlResponseString = firstTool?.function.arguments

    if (!sqlResponseString) {
      throw new EmptyResponseError()
    }

    const repairedJsonString = jsonrepair(sqlResponseString)

    const result: DebugSqlResult = JSON.parse(repairedJsonString)

    if (!result.sql) {
      throw new EmptySqlError()
    }

    return result
  } catch (error) {
    if (error instanceof Error && 'code' in error && error.code === 'context_length_exceeded') {
      throw new ContextLengthError()
    }
    throw error
  }
}

/**
 * Generates a snippet title based on the provided SQL.
 *
 * @returns A title and description for the SQL snippet.
 */
export async function titleSql(openai: OpenAI, sql: string) {
  const completionMessages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = [
    {
      role: 'user',
      content: sql,
    },
  ]

  try {
    const completionResponse = await openai.chat.completions.create({
      model: 'gpt-3.5-turbo-0125',
      messages: completionMessages,
      max_tokens: 1024,
      temperature: 0,
      tool_choice: {
        type: 'function',
        function: {
          name: completionFunctions.generateTitle.name,
        },
      },
      tools: [
        {
          type: 'function',
          function: completionFunctions.generateTitle,
        },
      ],
      stream: false,
    })

    const [firstChoice] = completionResponse.choices
    const [firstTool] = firstChoice.message?.tool_calls ?? []

    const sqlResponseString = firstTool?.function.arguments

    if (!sqlResponseString) {
      throw new EmptyResponseError()
    }

    const repairedJsonString = jsonrepair(sqlResponseString)

    const result: GenerateTitleResult = JSON.parse(repairedJsonString)

    return result
  } catch (error) {
    if (error instanceof Error && 'code' in error && error.code === 'context_length_exceeded') {
      throw new ContextLengthError()
    }
    throw error
  }
}
