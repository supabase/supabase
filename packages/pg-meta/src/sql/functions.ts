export const FUNCTIONS_SQL = /* SQL */ `
-- CTE with sane arg_modes, arg_names, and arg_types.
-- All three are always of the same length.
-- All three include all args, including OUT and TABLE args.
with functions as (
  select
    *,
    -- proargmodes is null when all arg modes are IN
    coalesce(
      p.proargmodes,
      array_fill('i'::text, array[cardinality(coalesce(p.proallargtypes, p.proargtypes))])
    ) as arg_modes,
    -- proargnames is null when all args are unnamed
    coalesce(
      p.proargnames,
      array_fill(''::text, array[cardinality(coalesce(p.proallargtypes, p.proargtypes))])
    ) as arg_names,
    -- proallargtypes is null when all arg modes are IN
    coalesce(p.proallargtypes, p.proargtypes) as arg_types,
    array_cat(
      array_fill(false, array[pronargs - pronargdefaults]),
      array_fill(true, array[pronargdefaults])) as arg_has_defaults
  from
    pg_proc as p
  where
    p.prokind = 'f'
)
select
  f.oid::int8 as id,
  n.nspname as schema,
  f.proname as name,
  l.lanname as language,
  case
    when l.lanname = 'internal' then ''
    else f.prosrc
  end as definition,
  case
    when l.lanname = 'internal' then f.prosrc
    else pg_get_functiondef(f.oid)
  end as complete_statement,
  coalesce(f_args.args, '[]') as args,
  pg_get_function_arguments(f.oid) as argument_types,
  pg_get_function_identity_arguments(f.oid) as identity_argument_types,
  f.prorettype::int8 as return_type_id,
  pg_get_function_result(f.oid) as return_type,
  nullif(rt.typrelid::int8, 0) as return_type_relation_id,
  f.proretset as is_set_returning_function,
  case
    when f.provolatile = 'i' then 'IMMUTABLE'
    when f.provolatile = 's' then 'STABLE'
    when f.provolatile = 'v' then 'VOLATILE'
  end as behavior,
  f.prosecdef as security_definer,
  f_config.config_params as config_params
from
  functions f
  left join pg_namespace n on f.pronamespace = n.oid
  left join pg_language l on f.prolang = l.oid
  left join pg_type rt on rt.oid = f.prorettype
  left join (
    select
      oid,
      jsonb_object_agg(param, value) filter (where param is not null) as config_params
    from
      (
        select
          oid,
          (string_to_array(unnest(proconfig), '='))[1] as param,
          (string_to_array(unnest(proconfig), '='))[2] as value
        from
          functions
      ) as t
    group by
      oid
  ) f_config on f_config.oid = f.oid
  left join (
    select
      oid,
      jsonb_agg(jsonb_build_object(
        'mode', t2.mode,
        'name', name,
        'type_id', type_id,
        'has_default', has_default
      )) as args
    from
      (
        select
          oid,
          unnest(arg_modes) as mode,
          unnest(arg_names) as name,
          unnest(arg_types)::int8 as type_id,
          unnest(arg_has_defaults) as has_default
        from
          functions
      ) as t1,
      lateral (
        select
          case
            when t1.mode = 'i' then 'in'
            when t1.mode = 'o' then 'out'
            when t1.mode = 'b' then 'inout'
            when t1.mode = 'v' then 'variadic'
            else 'table'
          end as mode
      ) as t2
    group by
      t1.oid
  ) f_args on f_args.oid = f.oid
`
