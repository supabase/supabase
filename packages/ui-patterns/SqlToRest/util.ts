import { HttpRequest, Statement, SupabaseJsQuery } from '@supabase/sql-to-rest'

export type BaseResult = {
  statement: Statement
}

export type HttpResult = BaseResult &
  HttpRequest & {
    type: 'http'
    language: 'http' | 'curl'
  }

export type SupabaseJsResult = BaseResult &
  SupabaseJsQuery & {
    type: 'supabase-js'
    language: 'js'
  }

export type ResultBundle = HttpResult | SupabaseJsResult
