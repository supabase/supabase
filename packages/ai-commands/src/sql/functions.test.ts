import { describe, expect, test } from '@jest/globals'
import { codeBlock } from 'common-tags'
import OpenAI from 'openai'
import { formatSql } from '../../test/util'
import { debugSql, titleSql } from './functions'

const openAiKey = process.env.OPENAI_API_KEY
const openai = new OpenAI({ apiKey: openAiKey })

describe('debug', () => {
  test.concurrent('fix order of operations', async () => {
    const { sql } = await debugSql(
      openai,
      'relation "departments" does not exist',
      codeBlock`
        create table employees (
          id bigint primary key generated always as identity,
          name text,
          email text,
          department_id bigint references departments (id)
        );

        create table departments (
          id bigint primary key generated always as identity,
          name text
        );
      `
    )

    expect(formatSql(sql)).toMatchSnapshot()
  })

  test.concurrent('fix typos', async () => {
    const { sql, solution } = await debugSql(
      openai,
      'syntax error at or near "fromm"',
      codeBlock`
        select * fromm employees;
      `
    )

    expect(solution).toBeDefined()
    expect(formatSql(sql)).toMatchSnapshot()
  })
})

describe('title', () => {
  test('title matches content', async () => {
    const { title, description } = await titleSql(
      openai,
      codeBlock`
        create table employees (
          id bigint primary key generated always as identity,
          name text,
          email text,
          department_id bigint references departments (id)
        );

        create table departments (
          id bigint primary key generated always as identity,
          name text
        );
      `
    )

    await expect(title).toMatchCriteria('relates to employees and departments')
    await expect(description).toMatchCriteria('describes employees and departments')
  })
})
