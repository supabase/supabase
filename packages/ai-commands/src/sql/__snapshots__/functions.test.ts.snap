// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug fix order of operations 1`] = `
"create table departments (
  id bigint primary key generated always as identity,
  name text
);

create table employees (
  id bigint primary key generated always as identity,
  name text,
  email text,
  department_id bigint references departments (id)
);"
`;

exports[`debug fix typos 1`] = `
"select
  *
from
  employees;"
`;
